packages/playwright-core/src/server/injected/ariaSnapshot.ts

```diff
@@ -0,0 +1,240 @@
+        normalizedChildren.push(child);
+      }
+    }
+    flushChildren(buffer, normalizedChildren);
+    ariaNode.children = normalizedChildren.length ? normalizedChildren : undefined;
+  };
+  visit(rootA11yNode);
+}
+
+const leafRoles = new Set([
+  'alert', 'blockquote', 'button', 'caption', 'checkbox', 'code', 'columnheader',
+  'definition', 'deletion', 'emphasis', 'generic', 'heading', 'img', 'insertion',
+  'link', 'menuitem', 'menuitemcheckbox', 'menuitemradio', 'meter', 'none', 'option',
+  'presentation', 'progressbar', 'radio', 'rowheader', 'scrollbar', 'searchbox', 'separator',
+  'slider', 'spinbutton', 'strong', 'subscript', 'superscript', 'switch', 'tab', 'term',
+  'textbox', 'time', 'tooltip'
+]);
+
+const normalizeWhitespace = (text: string) => text.replace(/[\s\n]+/g, ' ');
+
+function escapeRegex(string: string) {
+  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+}
+
+export function matchesAriaTree(injectedScript: InjectedScript, rootElement: Element, template: AriaTemplateNode): { matches: boolean, received: string } {
+  const root = generateAriaTree(injectedScript, rootElement);
+  const matches = nodeMatches(root, template);
+  return { matches, received: renderAriaTree(injectedScript, root) };
+}
+
+function matchesNode(node: AriaNode | string, template: AriaTemplateNode | AriaTemplateString, depth: number): boolean {
+  if (typeof node === 'string' && template.kind === 'string') {
+    const pattern = template.chunks.map(s => typeof s === 'string' ? escapeRegex(normalizeWhitespace(s)) : s.source);
+    return !!node.match(new RegExp(pattern.join('')));
+  }
+
+  if (typeof node === 'object' && template.kind === 'node') {
+    if (template.role && template.role !== node.role)
+      return false;
+    if (template.role && template.name !== node.name)
+      return false;
+    if (!containsList(node.children || [], template.children || [], depth))
+      return false;
+    return true;
+  }
+  return false;
+}
+
+function containsList(children: (AriaNode | string)[], template: (AriaTemplateNode | AriaTemplateString)[], depth: number): boolean {
+  if (template.length > children.length)
+    return false;
+  const cc = children.slice();
+  const tt = template.slice();
+  for (let t = tt.shift(); t; t = tt.shift()) {
+    let c = cc.shift();
+    while (c) {
+      if (matchesNode(c, t, depth + 1))
+        break;
+      c = cc.shift();
+    }
+    if (!c)
+      return false;
+  }
+  return !tt.length;
+}
+
+function nodeMatches(root: AriaNode, template: AriaTemplateNode): boolean {
+  const results: (AriaNode | string)[] = [];
+  const visit = (node: AriaNode | string): boolean => {
+    if (matchesNode(node, template, 0)) {
+      results.push(node);
+      return true;
+    }
+    if (typeof node === 'string')
+      return false;
+    for (const child of node.children || []) {
+      if (visit(child))
+        return true;
+    }
+    return false;
+  };
+  visit(root);
+  return !!results.length;
```

Comment:
It seems that we do not allow to skip intermediate nodes? For example,
```html
<ul><li><a>click me</a></li></ul>
```
and
```tsx
<x.list><x.link>click me</x.link></x.list>
```
Maybe we should allow that.

Reply:
Correct, we do not and yes, maybe we should do.